File: backend/src/main/java/de/codeministry/leadgen/ask/AdvertAnswer.java
/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * Copyright 2026 Marcello Muscara (codeministry)
 *
 * Licensed under the Apache License, Version 2.0. You may obtain a copy of the
 * License at http://www.apache.org/licenses/LICENSE-2.0
 */
package de.codeministry.leadgen.ask;

/**
 * What one advert says about one question.
 *
 * @param question the question that was asked, by its wire name.
 * @param stated   whether the advert says anything about it at all. <b>The distinction this
 *                 record exists for.</b> "The advert is silent on the rate" and "the advert
 *                 offers 85 EUR" are different answers, and a screen that renders a shrug the
 *                 same way it renders a fact is a screen that invites the reader to fill the
 *                 gap themselves. Same shape as an unscored offer, which is not a zero.
 * @param answer   one or two sentences, in the language of the advert, or null when
 *                 {@code stated} is false.
 * @param quote    the sentence from the advert the answer rests on, verbatim, or null. <b>A
 *                 claim without one is dropped rather than shown</b> — see {@link AdvertAsker}.
 * @param model    which model answered, so two answers that disagree can be told apart.
 */
public record AdvertAnswer(String question, boolean stated, String answer, String quote, String model) {

    /** The advert says nothing about this, which is an answer and not a failure. */
    public static AdvertAnswer silent(AdvertQuestion question, String model) {
        return new AdvertAnswer(question.key(), false, null, null, model);
    }
}

File: frontend/src/app/core/model/advert-answer.ts
/**
 * What one advert says about one bounded question.
 *
 * Mirrors `de.codeministry.leadgen.ask.AdvertAnswer`.
 *
 * @param stated whether the advert addresses the question at all. **The distinction this type
 *     exists for**, and the screen has to render the two differently: "the advert is silent on
 *     the rate" and "the advert offers 95 EUR" are different answers, and a shrug drawn like a
 *     fact invites the reader to fill the gap themselves.
 * @param answer one or two sentences in the advert's language, or null when it is silent.
 * @param quote the sentence from the advert the answer rests on, verbatim. The server drops any
 *     claim whose quote it could not find in the advert, so a rendered answer always has one —
 *     which is what makes it checkable by looking up.
 * @param model which model answered, so two answers that disagree can be told apart.
 */
export interface AdvertAnswer {
    readonly question: string;
    readonly stated: boolean;
    readonly answer: string | null;
    readonly quote: string | null;
    readonly model: string;
}

/**
 * The questions an advert answers. The server holds the same list as an enum and refuses
 * anything else, so this is a mirror and never the source.
 */
export const ADVERT_QUESTIONS = ['rate', 'client', 'onsite', 'onboarding', 'extension'] as const;

export type AdvertQuestion = (typeof ADVERT_QUESTIONS)[number];
